php-checks/src/main/java/org/sonar/php/checks/UseOfUninitializedVariableCheck.java

```diff
@@ -87,175 +97,239 @@
+    InitialDataCollector initialDataCollector = new InitialDataCollector();
+    tree.accept(initialDataCollector);
+    // Catch clauses do not appear in the CFG. We collect all exception variables in the function body and
+    // consider them as provided in the whole function body to avoid false positives.
+    providedVariables.addAll(initialDataCollector.exceptionVariables);
+
+    Map<CfgBlock, BlockSummary> predecessorsSummary = new HashMap<>();
+    cfgBlocks.forEach(
+      b -> predecessorsSummary.put(b, new BlockSummary(new HashSet<>(providedVariables), new HashSet<>(), false))
+    );
+    Deque<CfgBlock> workList = new ArrayDeque<>(cfgBlocks);
+
+    while (!workList.isEmpty()) {
+      CfgBlock block = workList.pop();
+
+      BlockSummary summary = getBlockSummary(block, predecessorsSummary.get(block));
+
+      for (CfgBlock successor : block.successors()) {
+        BlockSummary successorInit = predecessorsSummary.get(successor);
+        if (!successorInit.initializedFromPredecessors.containsAll(summary.initializedVariables())
+          || successorInit.scopeWasChanged != summary.scopeWasChanged) {
+          successorInit.initializedFromPredecessors.addAll(summary.initializedVariables());
+          successorInit.scopeWasChanged = successorInit.scopeWasChanged || summary.scopeWasChanged;
+          workList.add(successor);
+        }
+      }
+    }
 
-    final FunctionTree currentFunction;
+    Map<String, Set<Tree>> uninitializedVariableUses = new HashMap<>();
+    cfgBlocks.forEach(b -> checkBlock(b, predecessorsSummary.get(b))
+      .forEach((v, s) -> uninitializedVariableUses.computeIfAbsent(v, st -> new HashSet<>()).addAll(s)));
+
+    // static variables could be initialized after first usage. We only raise an issue on them when they were never
+    // initialized (i.e., not initialized by predecessors of the end block)
+    uninitializedVariableUses.entrySet().stream()
+      .filter(e -> !isInitializedStaticVariable(e.getKey(),
+        initialDataCollector.uninitializedStaticVariables,
+        predecessorsSummary.get(cfg.end())))
+      .forEach(e -> reportOnFirstTree(e.getValue()));
+  }
 
-    boolean trustVariables = true;
+  private static boolean isInitializedStaticVariable(String var, Set<String> uninitializedStaticVariables, BlockSummary endBlockSummary) {
+    return uninitializedStaticVariables.contains(var) && endBlockSummary.initializedFromPredecessors.contains(var);
+  }
 
-    Map<String, VariableIdentifierTree> firstVariableReadAccess = new HashMap<>();
+  private void reportOnFirstTree(Set<Tree> trees) {
+    trees.stream()
+      .min(Comparator.comparingInt(a -> ((PHPTree) a).getFirstToken().line())
+        .thenComparing(a -> ((PHPTree) a).getFirstToken().column()))
+      .ifPresent(t -> newIssue(t, MESSAGE));
+  }
+
+  private static boolean blockIsUnreachable(CfgBlock block, ControlFlowGraph cfg) {
+    return block.predecessors().isEmpty() && !block.equals(cfg.start());
```

Comment:
It's more subtle than that. A block is also unreachable if all its predecessors are unreachable.
```
function f($p) {
  return;
  // unreachable
  echo $a;
  if ($b) {
    // unreachable
  }
  // unreachable
  echo $c;
}
```